### PR TITLE
Update HVB Rewards Integration + Fix MovingAverage IsAlive Bug

### PR DESCRIPTION
**Pull request by KMFODA**
- [x] Update how the HVB scores are added to the moving average scores
- [x] Fix bugs in the get_scores api
- [x] Add in logging to the weights and averages API
- [x] Fix the stagnant moving averages error for dormant UIDs

### CODE REVIEW

File: neurons/validator/forward.py [hunk 1]
```diff
@@ -23,6 +27,58 @@
...
+                human_voting_bot_scores = {}
+
+                for inner_dict in human_voting_bot_round_scores.values():
+                    for key, value in inner_dict.items():
+                        if key in human_voting_bot_scores:
+                            human_voting_bot_scores[key] += value
+                        else:
+                            human_voting_bot_scores[key] = value
+                
+                human_voting_bot_scores = torch.tensor([human_voting_bot_scores[key] if key in human_voting_bot_scores.keys() else 0 for key in self.hotkeys]).to(self.device)
```
Comment (tavurth) [inline neurons/validator/forward.py:62]: I guess it can be nice to format with `black`?

`pip install black`

File: neurons/validator/forward.py [hunk 2]
```diff
@@ -247,92 +304,37 @@ def run_step(self, prompt, axons, uids, task_type="text_to_image", image=None):
     scattered_rewards: torch.FloatTensor = self.moving_averaged_scores.scatter(
         0, uids, rewards
     ).to(self.device)
+    
+    reward_i, reward_i_normalized =  self.human_voting_bot_reward_model.get_rewards(self.hotkeys)
+    scattered_rewards_adjusted = scattered_rewards + (self.human_voting_bot_weight*self.human_voting_bot_scores)
+    self.hvb_df =  pd.concat([self.hvb_df, pd.DataFrame({i: value.item() for i, value in enumerate(reward_i_normalized)}, index = [0])]).reset_index(drop=True)
```
Comment (tavurth) [inline neurons/validator/forward.py:310]: Also a black line cut

File: neurons/validator/reward.py [hunk 3]
```diff
@@ -349,6 +351,74 @@ def normalize_rewards(self, rewards: torch.FloatTensor) -> torch.FloatTensor:
...
+
+                elif (human_voting_scores.status_code != 200) and (attempt != max_retries):
+                    
+                    continue
+
+                else:
+                    
+                    human_voting_bot_round_scores = human_voting_scores.json()
+                    
+                    human_voting_bot_scores = {}
```
Comment (tavurth) [inline neurons/validator/reward.py:392]: We don't seem to use this?

File: neurons/validator/forward.py (hunk 1 above)
Comment (KMFODA) [reply neurons/validator/forward.py]: Ah good spot! I have the extension installed by default and it applies black formatting every time I save automatically. This must have been missed by that extension for some unknown reason. Just resolved.

File: neurons/validator/reward.py [hunk 4]
```diff
@@ -349,6 +351,79 @@ def normalize_rewards(self, rewards: torch.FloatTensor) -> torch.FloatTensor:
         return rewards
 
 
+class HumanValidationBotRewardModel(BaseRewardModel):
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/reward.py:354]: Change to HumanValidationRewardModel

File: neurons/validator/forward.py [hunk 5]
```diff
@@ -248,99 +250,69 @@ def run_step(self, prompt, axons, uids, task_type="text_to_image", image=None):
...
+            timeout=30,
+        )
+        if response.status_code != 200:
+            bt.logging.info("Error logging moving averages to the Averages API")
+        else:
+            bt.logging.info("Successfully logged moving averages to the Averages API")
+    except:
+        bt.logging.info("Error logging moving averages to the Averages API")
+
+    self.ma_df = pd.concat(
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/forward.py:300]: Are we using dfs just for testing purposes?

File: neurons/miners/StableMiner/base.py [hunk 6]
```diff
@@ -109,8 +109,11 @@ def start_axon(self):
             bt.axon(
                 wallet=self.wallet,
                 ip=bt.utils.networking.get_external_ip(),
-                external_ip=bt.utils.networking.get_external_ip(),
-                config=self.config
+                external_ip=self.config.axon.get(
```
Comment (Supreme-Emperor-Wang) [inline neurons/miners/StableMiner/base.py:112]: Should this be:
`external_ip=self.config.axon.get("external_ip") or bt.utils.networking.get_external_ip(),`

?

File: neurons/validator/reward.py [hunk 7]
```diff
@@ -349,6 +351,79 @@ def normalize_rewards(self, rewards: torch.FloatTensor) -> torch.FloatTensor:
...
+    @property
+    def name(self) -> str:
+        return RewardModelType.nsfw.value
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/reward.py:357]: This should return RewardModelType.**human**.value?

File: neurons/validator/reward.py [hunk 8]
```diff
@@ -349,6 +351,79 @@ def normalize_rewards(self, rewards: torch.FloatTensor) -> torch.FloatTensor:
...
+
+    def __init__(self, metagraph, api_url):
+        super().__init__()
+        self.device = "cuda"
+        self.human_voting_bot_scores = torch.zeros((metagraph.n)).to(self.device)
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/reward.py:362]: Remove the bot component since HVP also uses this

File: neurons/validator/reward.py [hunk 9]
```diff
@@ -349,6 +351,79 @@ def normalize_rewards(self, rewards: torch.FloatTensor) -> torch.FloatTensor:
...
+        self.api_url = api_url
+
+    def get_rewards(self, hotkeys) -> torch.FloatTensor:
+        max_retries = 3
+        backoff = 2
+
+        bt.logging.info("Extracting human votes")
+
+        human_voting_bot_scores = None
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/reward.py:371]: Same here

Comment (Supreme-Emperor-Wang) [reply neurons/validator/reward.py:371]: We also have a variable called human_voting_scores though, so please rename to create some distinction. Maybe human_voting_scores_raw/processed?

File: neurons/validator/reward.py [hunk 10]
```diff
@@ -349,6 +351,79 @@ def normalize_rewards(self, rewards: torch.FloatTensor) -> torch.FloatTensor:
...
+                    attempt != max_retries
+                ):
+                    continue
+
+                else:
+                    human_voting_bot_round_scores = human_voting_scores.json()
+
+                    human_voting_bot_scores = {}
+
+                    for inner_dict in human_voting_bot_round_scores.values():
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/reward.py:396]: Remove bot

File: neurons/validator/reward.py [hunk 11]
```diff
@@ -349,6 +351,79 @@ def normalize_rewards(self, rewards: torch.FloatTensor) -> torch.FloatTensor:
...
+                break
+
+        if human_voting_bot_scores is not None:
+            for index, hotkey in enumerate(hotkeys):
+                if hotkey in human_voting_bot_scores.keys():
+                    self.human_voting_bot_scores[index] = human_voting_bot_scores[
+                        hotkey
+                    ]
+
+        human_voting_bot_scores_normalised = (
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/reward.py:420]: remove bot here also

File: neurons/validator/utils.py [hunk 12]
```diff
@@ -57,26 +57,34 @@ def ttl_get_block(self) -> int:
...
+            timeout=self.async_timeout,
+        )
         if response.is_success:
             response_times.append(time.perf_counter() - t1)
+            self.isalive_dict[uid] = 0
             return True
         else:
-
             try:
+                self.isalive_dict[uid] += 1
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/utils.py:72]: Rename variable to indicate that this is tracking failed isalive attempts. Maybe` isalive_failed_dict`

File: neurons/validator/validator.py [hunk 13]
```diff
@@ -242,6 +262,10 @@ def __init__(self):
         self.hotkey_whitelist = set()
         self.coldkey_whitelist = set()
 
+        # Init IsAlive counter
+        self.isalive_threshold = 3
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/validator.py:266]: I think this should be higher. Let's say 8-10?

File: neurons/validator/validator.py [hunk 14]
```diff
@@ -257,21 +281,60 @@ def __init__(self):
...
         try:
-            self.miner_query_history_fail_count = {self.metagraph.axons[uid].hotkey:0 for uid in range(self.metagraph.n.item())}
+            self.miner_query_history_fail_count = {
+                self.metagraph.axons[uid].hotkey: 0
+                for uid in range(self.metagraph.n.item())
+            }
         except:
             pass
 
+        # Load Moving Average and Weight Dataframe
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/validator.py:308]: Is this just for testing?

File: neurons/validator/weights.py [hunk 15]
```diff
@@ -29,6 +31,36 @@ def set_weights(self):
...
+            timeout=30,
+        )
+        if response.status_code != 200:
+            bt.logging.info("Error logging weights to the Weights API")
+        else:
+            bt.logging.info("Successfully logged weights to the Weights API")
+    except:
+        bt.logging.info("Error logging weights to the Weights API")
+
+    self.weights_df = pd.concat(
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/weights.py:56]: Testing?

**Review by Supreme-Emperor-Wang — CHANGES_REQUESTED** [review]

File: neurons/validator/utils.py (hunk 12 above)
Comment (KMFODA) [reply neurons/validator/utils.py:72]: good suggestion, changed to `isalive_failed_count`

File: neurons/validator/validator.py (hunk 14 above)
Comment (KMFODA) [reply neurons/validator/validator.py:308]: It was also for logging purposes for each validator but now that we have the moving_averages live chart we can remove this.
